ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java

```diff
@@ -0,0 +1,476 @@
+
+import com.github.ambry.clustermap.MockClusterMap;
+import com.github.ambry.clustermap.PartitionId;
+import com.github.ambry.commons.BlobId;
+import com.github.ambry.commons.ByteBufferAsyncWritableChannel;
+import com.github.ambry.commons.ByteBufferReadableStreamChannel;
+import com.github.ambry.config.RouterConfig;
+import com.github.ambry.config.VerifiableProperties;
+import com.github.ambry.coordinator.AmbryCoordinator;
+import com.github.ambry.messageformat.BlobInfo;
+import com.github.ambry.messageformat.BlobProperties;
+import com.github.ambry.router.Callback;
+import com.github.ambry.router.CoordinatorBackedRouter;
+import com.github.ambry.router.CoordinatorBackedRouterMetrics;
+import com.github.ambry.router.NonBlockingRouterFactory;
+import com.github.ambry.router.ReadableStreamChannel;
+import com.github.ambry.router.Router;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Queue;
+import java.util.Random;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Future;
+import org.junit.Assert;
+
+
+class RouterServerTestFramework {
+  static final int CHUNK_SIZE = 1024 * 1024;
+  private final MockClusterMap clusterMap;
+  private final MockNotificationSystem notificationSystem;
+  private final Router nonBlockingRouter;
+  private final Router coordinatorBackedRouter;
+
+  RouterServerTestFramework(Properties routerProps, MockCluster cluster, MockNotificationSystem notificationSystem)
+      throws Exception {
+    this.clusterMap = cluster.getClusterMap();
+    this.notificationSystem = notificationSystem;
+
+    VerifiableProperties routerVerifiableProps = new VerifiableProperties(routerProps);
+    this.nonBlockingRouter =
+        new NonBlockingRouterFactory(routerVerifiableProps, clusterMap, notificationSystem).getRouter();
+
+    RouterConfig routerConfig = new RouterConfig(routerVerifiableProps);
+    CoordinatorBackedRouterMetrics coordinatorBackedRouterMetrics =
+        new CoordinatorBackedRouterMetrics(clusterMap.getMetricRegistry());
+    AmbryCoordinator coordinator = new HelperCoordinator(routerVerifiableProps, clusterMap);
+    this.coordinatorBackedRouter =
+        new CoordinatorBackedRouter(routerConfig, coordinatorBackedRouterMetrics, coordinator);
+  }
+
+  void cleanup()
+      throws IOException {
+    if (nonBlockingRouter != null) {
+      nonBlockingRouter.close();
+    }
+    if (coordinatorBackedRouter != null) {
+      coordinatorBackedRouter.close();
+    }
+  }
+
+  void checkOperationChains(List<OperationInfo> opInfos)
+      throws Exception {
+    Map<PartitionId, Integer> partitionCount = new HashMap<>();
+    double blobsPut = 0;
+    for (OperationInfo opInfo : opInfos) {
+      opInfo.latch.await();
+      synchronized (opInfo.futures) {
+        for (TestFuture future : opInfo.futures) {
+          future.check();
```

Comment:
Suggest rename to `testFuture` just to be very specific. Since you have `future` as well for the router operation result. Correspondingly, ln. 385 make some change.